### ISSUE TITLE
[sfm/clustering] Add a scene clustering approach

### ISSUE THREAD

**Issue by pmoulon** [openMVG/openMVG#857]
Add an implementation of "Dominant set clustering" introduced by Maura, Massimo et al. at ETHZ Zurich Computer Vision Lab.

**Reference papers are:**
Mauro, Massimo, et al. ["An integer linear programming model for view selection on overlapping camera clusters."](http://www.vision.ee.ethz.ch/~rhayko/paper/3dv2014_mauro_joint_selection_clustering.pdf) 2014 2nd International Conference on 3D Vision. Vol. 1. IEEE, 2014.

Mauro, Massimo, et al. ["Overlapping camera clustering through dominant sets for scalable 3D reconstruction."](http://www.vision.ee.ethz.ch/~rhayko/paper/bmvc2013_mauro_view_clustering.pdf) 2013 British Machine Vision Conference.

- [x] merge to develop
- [x] remove the nanoflann dependency

**Comment by pmoulon** [comment]
Here some clustering results:

1229 cameras => automatic clustering -> 50 group:
![brinon_sur_sauldre_eglise00](https://cloud.githubusercontent.com/assets/3235050/24827515/aac25cae-1c4b-11e7-9544-510af1135763.png)

2529 cameras => automatic clustering -> 106 group:
![egliselye00](https://cloud.githubusercontent.com/assets/3235050/24827520/c7a597c8-1c4b-11e7-9af0-495691b773b0.png)

1464 cameras => automatic clustering -> 61 group:
![meusnes01](https://cloud.githubusercontent.com/assets/3235050/24827521/d82c6f04-1c4b-11e7-9f8d-6480e576c690.png)

152 cameras => automatic clustering -> 6 group:
![eglisesaintlys00](https://cloud.githubusercontent.com/assets/3235050/24827525/e87edffe-1c4b-11e7-8d67-63f46d4fbba2.png)

422 cameras => automatic clustering -> 18 group:
![saintsulspice200](https://cloud.githubusercontent.com/assets/3235050/24827538/028116ba-1c4c-11e7-8cf5-84d42776a863.png)

Then you can process each cluster with the MVS framework of your choice and create stunning point clouds on a basic computer:

Here the point clouds produced by OpenMVS:
![snapshot_ptcloud01](https://cloud.githubusercontent.com/assets/3235050/24827562/8a477666-1c4c-11e7-9c0b-9d9dce299c9f.png)

Here the point clouds produced by PMVS:
![snapshot02](https://cloud.githubusercontent.com/assets/3235050/24827689/198a1106-1c4f-11e7-90c2-93824967e0f0.png)

**Comment by bigakoz** [comment]
Pierre,

I have downloaded this WIP and compiling now.

How do the settings work for the command line arguments
           << "[-l|--cluster_size_lower_bound] lower bound to cluster size\n"
           << "[-u|--cluster_size_upper_bound] upper bound to cluster size\n"
           << "[-v|--voxel_grid_size] voxel grid size\n"

I have  a large data set of high resolution photography (750 images) which is to big for my computer configuration.

If I could break this up into groups of 100 images each would be great.

The imagery is all oblique and unordered so it is not possible to organize easily into groups. I am hoping this version does the trick.

Regards,

Alan

**Comment by pmoulon** [comment]
@bigakoz Keep us updated!

Avoid to touch to the parameter -v.

Just touch to the `--cluster_size_upper_bound` parameter is sufficient for most of the project.

You can try `--cluster_size_upper_bound 100` in order divide by 7 the complexity of the MVS project.

You can try `--cluster_size_upper_bound 50` in order divide by 14 the complexity of the MVS project.

**Comment by bigakoz** [comment]
Pierre,

I tried this but it isn't exactly what I wanted.

In summary.
1. I have 750 images from aerial photography.
2. The acquisition as 4 cameras mounted obliquely.
3. The images are unordered and almost impossible to organize into sub blocks
2. I have run a bundle on all the data.
3. Data is to big to run in openMVE and Textrecon for building a 3D model.
4. I want to create sub blocks to run the openMVE in parts.

I am writing a small function to identify structure and poses that fall within a user defined block.
I am slowly becoming familiar with the sfm_data formats.
I am able to build a vector that identifies the poses that I need for each block.

My next step is to re-create some sfm_data.bin files with the subset data.
I assume it is a simple process to delete from the pose and structure structures the unused parts and save to a new file name.

Alan

**Comment by pmoulon** [comment]
> My next step is to re-create some sfm_data.bin files with the subset data.
I assume it is a simple process to delete from the pose and structure structures the unused parts and save to a new file name.

You're right.

I would be curious to see the output of `main_ComputeClusters` on your dataset.
It should produce something reasonable.

**Comment by bigakoz** [comment]
Pierre,

I have managed to get clusters of my data by hacking the clustering code.
I didn't need anything as sophisticated as what was in the source code. 
I just broke the model up into rows and collumns (in my example a 8 x 8 grid) and filled each cluster by testing whether the structure points fell into each grid zone.

I deleted the original cluster output but can rerun if you are interested but the result from  my data resulted in a patch work of images that weren't a single model that wasn't what i was after.

There  was other issues with my data as well that made this complex. The OpenMVE pipeline uses floats which means it couldn't handle ECEF numbers which are large. I had to truncate the coordinates before running the bundle and inputting into the openMVE pipeline and then restore to proper ECEF at the end. So was a bit of mucking around.

**Comment by pmoulon** [comment]
@bigakoz If you want you can contribute a new binary that will normalize a sfm_data scene.

**Comment by bigakoz** [comment]
Pierre,
I can share but I have currently only done this for a scene in ECEF coordinates. So is probably not a robust general solution
i.e
1. I convert each point to lat,longitude and compare whether it is inside my 2D polygon.

To make this robust and applicable for data that isn't on a geodetic coordinate system I probably need sub-blocks that are a cube (rather than a rectangular block). Do you know of a simple algorithm that I can use ? I know of a simple one for texting whether a point is inside a polygon but have not searched for one that is inside a 3D structure.

ALan

**Comment by bigakoz** [comment]
Ignore my last question.

I realize I am using a cube (the axis are parallel) so is trivial to see if  a point is within a cube.

I will change my program to work with cubes (rather than a square).

I have worked on something else that may be of use with the community.
1. the openMVS pipeline doesn't work with large coordinates (ECEF) because it uses 'floats' rather than doubles.
2. A ortho generation program I have found also won't work if the coordinates are not in east,north or lat,long as it generates the orthophoto purely in the horizontal dimensions.

In order to handle this I have had to convert ECEF coordinates into a truncated coordinate system.
- rotate ECEF into a local coordinate system (e,n,h) and the truncating the coordinates.
- This can then be transformed back to ECEF (or whatever projection is required on completion of processing)

**Comment by cdcseacave** [comment]
@pmoulon great looking scene, where it is? could you share the photos? https://cloud.githubusercontent.com/assets/3235050/24827689/198a1106-1c4f-11e7-90c2-93824967e0f0.png

**Comment by pmoulon** [comment]
Hi @cdcseacave 
This is a `Eglise a caquetoire` in Brinon sur Sauldre France. It show a nice wood structure where people was used to sit and chat (explaining the french name `caquetoire`).

Will try to see if I can share the images. The dataset was quite big ;-)

**Comment by ManishSahu53** [comment]
@bigakoz can you share your clustering code which is used to divide into 8x8 grid? I am also working on aerial images and facing similar problem

**Comment by BigAloz** [comment]
Manish,

Sure, I am not doing anything fancy.

A. I supply a block of my area of interest.
B. I go through the Landmark datastructure and find points within my block.
C. I then create a array with Views that use the landmark points
D. Then write a new sfm_data.bin file using the reduced Views.

How do I post my source on this site ? or should I just email to you

**Comment by ManishSahu53** [comment]
@bigakoz ,
you can email or upload here too. My email is manish.sahu.civ13@iitbhu.ac.in. 
I will try this grid on my dataset. Then update its status here. Thank you.

**Comment by pmoulon** [comment]
Why not try to work on making a OpenMVG application that cluster:
 - a sfm_data with only view with GPS priors to overlapping blocks.
 - a sfm_data with geometry (poses + landmark) to overlapping cells.

**Comment by BigAloz** [comment]
Pierre, Manish,

We may not be on the same page here.

The reason I am clustering or down sampling is for the dense reconstruction. I break my project into sub blocks for the dense reconstruction.

So I down sample after doing the bundle adjustment to get clusters that are manageable sizes for the dense reconstruction as this is a very slow part of the pipeline.

I think Manish may want to do this prior to running the bundle adjustment. I haven't looked at this as I have found the bundle adjustment is not the road-block in the pipeline.

**Comment by ManishSahu53** [comment]
@BigAloz ,
I want to do dense reconstruction only. My dataset is large so I want to cluster them just like yours example 8x8 grid.

**Comment by BigAloz** [comment]
Thats fine. 

The method I used works OK for the data I have (aerial oblique) and made life much easier when doing the dense reconstruction

**Comment by pmoulon** [comment]
@ManishSahu53 @bigakoz I just meant that it would be sweet to see you contribute back to the project so that every user facing this could use your tool ;-)
And the community could even perform a code review of your tool and help to improve it in the future ;-)

**Comment by BigAloz** [comment]
I am more than happy to contribute but not sure how I submit code

**Comment by pmoulon** [comment]
You can submit new features by doing a Pull Request (PR). See https://github.com/openMVG/openMVG/blob/master/CONTRIBUTING.md

**Comment by ManishSahu53** [comment]
I have tested it in my data. Works good. I have added a bit extra in it. I have created a shp file with grids in it. The program takes that shp file as input and automatically cluster the sfm.bin file according to grid. Although I am getting points outside the grid too. Not sure If this is intentional by @BigAloz  but its working fine.

[SfM_Global_mvs.txt](https://github.com/openMVG/openMVG/files/1430214/SfM_Global_mvs.txt)
